### DISCUSSION TITLE
Automatically Filename Generation

### DISCUSSION THREAD

**Discussion by NatoNathan** [tinacms/tinacms#2561] [Ideas]
# Automatically Filename Generation
I have started to use TinaCMS but I happen to have a use case that currently is not supported, but I think would be of benefit to a lot of users, especially those with non-tech savvy’s editors.

It would be nice to be able to have some automatic generation of filenames. Potentially in the way of a callback or a hook.

## Callback idea

This could be a callback on `<TinaCMS/>` or on `documentCreatorCallback`, witch could the newly generated filename.

Maybe it could look something like
```js
import { uuid } from 'uuidv4';
...
<TinaCMS
  documentCreatorCallback={{
    onAddDocument((args)=> {
      const name = uuid();
      return {
        filename:`${name}.${args.format}`,
        initialData:{}
      }
    },
  }}
...
>
```
or
```js
...
<TinaCMS
  documentCreatorCallback={{
    onAddDocument((args)=> {
      const name = args.name.toLowerCase().replace(' ', '-');
      return {
        filename:`${name}.${args.format}`,
        initialData:{}
      }
    },
  }}
...
>
```

## Hook idea

Some users may want to display a Add new post button to editors in edit mode. A hook could be nice to make a new document with a generated name and some default values, and then navigate to the new post and open the edit panel.

```jsx
const BlogsPage =(props) => {
  const {addDoc} = useAddDocument();
  const addBlog = (e) => {
    const name = uuid();
    return {
      filename:`${name}.${args.format}`,
      initialData:{}
    }
  };
  return (
    <div>
    {/* ...BlogCards */}
    <button onClick={addBlog} />
    </div>
  )
}
```

**Comment by konsumer** [comment]
I could also see schema field-options being really useful, and then you could output whatever you want.

```js
collections: [
    {
      label: 'Page Content',
      name: 'page',
      path: 'content/page',
      format: 'mdx',
      fields: [
        {
          label: 'FILENAME',
          name: 'filename',
          type: 'string',
          isFilename: true
          component: MyReallyFancyFilenameComponent,
          parse(value, name, field) {},
          format(value, name, field) {}
        },
        {
          label: 'URL Slug',
          name: 'slug',
          type: 'string'
        },
       {
          label: 'Date',
          name: 'date',
          type: 'datetime'
        },
      ]
    }
]
```

I think the `component`/`parse`/`format` just needs access to the other values, and this could enable things like deriving filename from `slug` + `date` and keeping it hidden from the UI (like `MyReallyFancyFilenameComponent` returns null.)

I think your idea of a function could work here, too, without a full field.

-  if `filename` is set on collection, don't show it in UI.
- if `filename` is a function, run it (with other form values) to derive the value

**Comment by danielknell** [comment]
I'm currently bodging this with

```ts
        {
          type: "string",
          label: "Title",
          name: "title",
          isTitle: true,
          required: true,
          ui: {
            validate: (value, allValues, meta, field) => {
              allValues.filename = slugify(value);
            },
          },
        },
```

or

```ts
        {
          type: "string",
          label: "Title",
          name: "title",
          isTitle: true,
          required: true,
          ui: {
            validate: (value, allValues, meta, field) => {
              if (allValues.date) {
                allValues.filename = slugifyPost(allValues.date, value);
              }
            },
          },
        },
        {
          type: "datetime",
          label: "Date",
          name: "date",
          required: true,
          ui: {
            component: "date",
            dateFormat: "YYYY-MM-DD",
            timeFormat: "HH:mm:ss",
            validate: (value, allValues, meta, field) => {
              if (allValues.title) {
                allValues.filename = slugifyPost(value, allValues.title);
              }
            },
          },
        },
```

but a less hacky way would be great, even more so with the option to make the default `filename` field  an `<input type="hidden">` widget.

**Comment by konsumer** [comment]
@danielknell Thanks! I will try that.

So more on filename:

in tinacms/CollectionCreatePage has a pretty normal-looking form-schema:

```js
  const form = useMemo(() => {
    return new Form({
      id: 'create-form',
      label: 'form',
      fields: [
        {
          name: 'filename',
          label: 'Filename',
          component: 'text',
          description: (
            <span>
              A unique filename for the content.
              <br />
              Examples: <code>My_Document</code>, <code>My_Document.en</code>,{' '}
              <code>sub-folder/My_Document</code>
            </span>
          ),
          placeholder: `My_Document`,
          validate: (value, allValues, meta) => {
            if (!value) {
              if (meta.dirty) {
                return 'Required'
              }
              return true
            }

            const isValid = /^[_a-zA-Z0-9][\.\-_\/a-zA-Z0-9]*$/.test(value)
            if (value && !isValid) {
              return 'Must begin with a-z, A-Z, 0-9, or _ and contain only a-z, A-Z, 0-9, -, _, ., or /.'
            }
          },
        },
        ...(formInfo.fields as any),
      ],
      onSubmit: async (values) => {
        try {
          await createDocument(cms, collection, template, mutationInfo, values)
          cms.alerts.success('Document created!')
          navigate(`/collections/${collection.name}`)
        } catch (error) {
          console.error(error)

          throw new Error(
            `[${error.name}] CreateDocument failed: ${error.message}`
          )
        }
      },
    })
  }, [cms, collection, mutationInfo])
```

For advanced usecases like automating the filename, if I could override that form, it would do the same thing as exposing `filename` in the schema, for me. Is there an existing way to inject my own form here?

**Comment by konsumer** [comment]
@danielknell I tried your solution, and it mostly does what I need in this case. I'd like users to be able to override the filename, if they want to, and also I want to be able to change the wording in filename field (so it matches more how we use `slug` in forestry.)

**Reply by danielknell** [reply]
hacks be dirty, it was enough to unblock me for now though so thought I would share. :o)

**Comment by konsumer** [comment]
In fact, if I could just hide the built-in filename field, I could override it's value with another field named filename:

```js
{
  type: 'string',
  label: 'URL Slug',
  name: 'filename',
  description: 'A unique URL for your blog-post',
  required: true
}
```
![2022-08-24 13 22 28](https://user-images.githubusercontent.com/83857/186516025-dae8eda1-4838-4ff5-b5b2-e2b70a0f760d.gif)

Reading [this](https://tina.io/blog/add-and-delete-files/) it seems like I might be able to create a plugin for all this. I am going to try that next.

**Reply by danielknell** [reply]
```css
div[data-test="form:create-form"] .relative.mb-5.last\:mb-0:first-child {
    display: none !important;
}
```
how far into the filth are you willing to wade? 🤡

**Reply by konsumer** [reply]
It has another issue with the IDs for `filename` on 2 items being the same (errors in console) but I was totally thinking of something similar.

**Comment by konsumer** [comment]
As a total hack, I followed suggestion above with css:

```css
/* hide filename, so we can override it on "create" form. */
div[data-test="form:create-form"] .relative.mb-5.last\:mb-0:first-child {
    display: none !important;
}
```

and added this to my schema:

```js
// above schema
let pageTouched = false

{
  type: 'string',
  label: 'Title',
  name: 'title',
  isTitle: true,
  required: true,
  ui: {
    validate: (value, allValues, meta, field) => {
      if (value) {
        if (!pageTouched && document.location.hash === '#/collections/page/new') {
          allValues.filename = slugify(value, { strict: true, lower: true })
          allValues.slug = allValues.filename
        }
      }
    }
  }
},
{
  type: 'string',
  label: 'URL Slug',
  name: 'slug',
  ui: {
    validate: (value, allValues, meta, field) => {
      if (document.location.hash === '#/collections/page/new') {
        allValues.filename = value
      }
    },
    parse (value) {
      pageTouched = true
      return value
    }
  }
}
```

So only on the create form it sets filename, and there are no duplicate fields. This will allow users to edit the `slug` (which is what I will use to find a URL) and the filename will be loosely based on the initial slug they chose. Total hack, but it does accomplish what I was trying to do.

I think main consideration, is now I will need to always do this hackery on filename, since it will never show on any content-form, but that's ok with me.

**Comment by L-A** [comment]
Some input from our different solution: I ended up building a small `prebuild.js` file that parses all post files and renames them based on a field in their front matter (`slug` in our case). With a `prebuild` command added to `package.json`, it gets picked up automatically when running `build`.

While waiting for a way to do this straight in Tina, this lets any CI build the site right without getting in editors' way.

**Reply by L-A** [reply]
Update! After checking in with how Tina Cloud works, for this to be production-safe, you'd need for your CI/CD to actually commit  these changes if there are any, because Tina Cloud relies on the repo _before_ prebuild. This means you could introduce discrepancies between the two by changing filenames during a `prebuild` command. 😬

**Comment by jamespohalloran** [comment]
Going to toss out another potential solution we could implement on our end, that would save having to dive in the low level Tina code.

What if a collection could allow for a function like:

```
slugify((form): string => {
// ...
})
```

This function would generate the filename **only on document creation**, based on values from your form. 

If the user doesn't provide a slugify function, we'd provide an under the default using some magic to guess the filename:

```
slugify((form): string => {
	// ...
	
	return toFilename(isTitleField
	|| form.values["title"]
	|| form.values["name"]
	|| form.values[0] // Iuno about this, maybe the first string?
	|| (new Date()).toString()
	)
})
```

The existing filename field could be removed from the "addDocument" modal (or replaced with a readonly label like in @konsumer 's example: https://github.com/tinacms/tinacms/discussions/2561#discussioncomment-3469044)

<img width="964" alt="Screen Shot 2022-08-31 at 10 39 38 AM" src="https://user-images.githubusercontent.com/3323181/187694132-a67c34d6-95a6-41f4-975f-226511673153.png">

**Reply by NatoNathan** [reply]
A filename generation function would work but `slugify` is probably the wrong name.
`generateFilename()` or `filenameGenerator()` would be more accurate

**Reply by konsumer** [reply]
@NatoNathan In my example, I use "slugify" is the name of a library that does this for you: https://www.npmjs.com/package/slugify but I agree, maybe a better name, if it's to become a tina field API thing. Like I said, it worked fine to hide filename with CSS, and do it the way I did it.

**Reply by konsumer** [reply]
For me, simply having a `hidden` option, that would allow us to use validate/parse with a built-in field would solve this the best, I think. As it is, I can hide other things by setting their `component` to `() => null`, but since `filename` is built-in, there is no way to override it.

**Reply by NatoNathan** [reply]
@konsumer I get that, but @jamespohalloran was suggesting to call the filename generation callback `slugify()`,

Actually, this is another reason for not using `slugify()` as the name, it leads to ambiguity.

**Reply by danielknell** [reply]
tbh, if the filename widget was specifiable it would solve all this? want to rename it to slug? change the label. want to automate it completely? define a "hidden" widget that generates a `<input type="hidden" />` and have at it with the validators... and any other customisation hooks fields provide for free...

could maybe do it by checking if `filename` is defined in the schema and special casing it?

**Reply by konsumer** [reply]
Yep, we are sort of talking about the same thing here: https://github.com/tinacms/tinacms/issues/2706 I think `filename` should just work like other fields, with `isFilename` or something to clearly mark it as a "special field" (similar to `isTitle` or `isBody`) but otherwise work just like any other field.

**Comment by jeffsee55** [comment]
FYI this is a WIP [here](https://github.com/tinacms/tinacms/pull/3230)

**Comment by jeffsee55** [comment]
This has been released https://tina.io/docs/extending-tina/filename-customization/

Please give it a try and let us know how you like it!
